client.go

```diff
@@ -189,34 +189,100 @@ func (c *Client) exchange(m *Msg, a string) (r *Msg, rtt time.Duration, err erro
 // If the received message contains a TSIG record the transaction
 // signature is verified.
 func (co *Conn) ReadMsg() (*Msg, error) {
-	var p []byte
+	p, err := co.ReadMsgBytes(nil)
+	if err != nil {
+		return nil, err
+	}
+
 	m := new(Msg)
-	if _, ok := co.Conn.(*net.TCPConn); ok {
-		p = make([]byte, MaxMsgSize)
+	if err := m.Unpack(p); err != nil {
+		return nil, err
+	}
+	if t := m.IsTsig(); t != nil {
+		if _, ok := co.TsigSecret[t.Hdr.Name]; !ok {
+			return m, ErrSecret
+		}
+		// Need to work on the original message p, as that was used to calculate the tsig.
+		err = TsigVerify(p, co.TsigSecret[t.Hdr.Name], co.tsigRequestMAC, false)
+	}
+	return m, err
+}
+
+// ReadMsgBytes reads DNS packet, parses and fills wire header (passing nil
+// would cause skipping that). Returns message in byte format to parse with
+// Msg.Unpack later on.
+//
+// Note that this function would not be able to report TSIG error or
+// check it got actual DNS payload.
+func (co *Conn) ReadMsgBytes(hdr *Header) ([]byte, error) {
```

Comment:
[ Quoting <notifications@github.com> in "Re: [dns] Added function for lazy m..." ]
>> +// Note that this function would not be able to report TSIG error or
>> +// check it got actual DNS payload.
>> +func (co *Conn) ReadMsgBytes(hdr *Header) ([]byte, error) {
>
>Once you read from UDP rest is discarded. You could not really read again there so you end up buffering and faking ability to read which is slow.
>
>http://golang.org/src/net/udpsock_posix.go?s=1547:1620#L53
>http://golang.org/src/net/fd_unix.go#L256
>
>then from rcvfrom(2):
>
>    All three routines return the length of the message on successful completion. If a message
>    is too long to fit in the supplied buffer, excess bytes may be discarded depending on the
>    type of socket the message is received from.
>

Grrrr :( So what would be the win then, this is the typical space/time trade
off? So the allocation of memory still happens, the only thing that you shortcut
is the potentially slowish parsing of the rest of the packet. I rather spend
cycles on making that faster, or fail ealier.

/Miek

--
Miek Gieben